winch/codegen/src/trampoline.rs

```diff
@@ -0,0 +1,163 @@
+use crate::{
+    abi::{align_to, ABIArg, ABIResult, ABI},
+    codegen::call::calculate_frame_adjustment,
+    masm::{CallKind, MacroAssembler, OperandSize, RegImm},
+    reg::Reg,
+};
+use std::mem;
+use wasmparser::{FuncType, ValType};
+
+/// A trampoline to provide interopt between different calling conventions.
+pub(crate) struct Trampoline<'a, A, M>
+where
+    A: ABI,
+    M: MacroAssembler,
+{
+    /// The macro assembler.
+    masm: &'a mut M,
+    /// The ABI.
+    abi: &'a A,
+    /// A scratch register.
+    scratch_1: Reg,
+    /// A second scratch register. This can be allocatable for the callee.
+    scratch_2: Reg,
+}
+
+impl<'a, A, M> Trampoline<'a, A, M>
+where
+    A: ABI,
+    M: MacroAssembler,
+{
+    /// Create a new trampoline.
+    pub fn new(masm: &'a mut M, abi: &'a A, scratch_1: Reg, scratch_2: Reg) -> Self {
+        Self {
+            masm,
+            abi,
+            scratch_1,
+            scratch_2,
+        }
+    }
+
+    /// Emit the host to wasm trampoline.
+    pub fn emit_host_to_wasm(&mut self, ty: &FuncType) {
+        // The host to wasm trampoline is currently hard coded (see vmcontext.rs in the
+        // wasmtime-runtime crate, VMTrampoline)
+        let trampoline_ty = FuncType::new(
+            vec![ValType::I64, ValType::I64, ValType::I64, ValType::I64],
+            vec![],
+        );
+        let trampoline_sig = self.abi.sig(&trampoline_ty);
```

Comment:
I think I might just leave a note for this one and address later on. I think this will be a bit of a bigger change to plumb it through properly, and this PR is getting big as it is 😅

Reply:
Yeah no worries, just wanted to note that so that you can explicitly call it out so that we can follow up on this.

Reply:
This is going to be part of the _ABI compliance_ note that I added [here](https://github.com/KevinRizzoTO/wasmtime/pull/7#issuecomment-1487767204)

Reply:
Thanks! Yeah I'll make sure to add this in the main PR